homeassistant/components/switch/netio.py

```diff
@@ -0,0 +1,229 @@
+
+Device = namedtuple('device', ['netio', 'entities'])
+DEVICES = {}
+ATTR_START_DATE = 'start_date'
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+
+def setup_platform(hass, config, add_devices_callback, discovery_info=None):
+    """Configure the netio platform."""
+    from pynetio import Netio
+
+    if validate_config({"conf": config}, {"conf": [CONF_OUTLETS,
+                                                   CONF_HOST]}, _LOGGER):
+        if len(DEVICES) == 0:
+            hass.wsgi.register_view(NetioApiView)
+
+        dev = Netio(config[CONF_HOST],
+                    config.get(CONF_PORT, DEFAULT_PORT),
+                    config.get(CONF_USERNAME, DEFAULT_USERNAME),
+                    config.get(CONF_PASSWORD, DEFAULT_USERNAME))
+        # if not dev.telnet:
+        #     return False
+
+        DEVICES[config[CONF_HOST]] = Device(dev, [])
+
+        for key in config[CONF_OUTLETS]:
+            switch = NetioSwitch(DEVICES[config[CONF_HOST]].netio, key,
+                                 config[CONF_OUTLETS][key])
+            DEVICES[config[CONF_HOST]].entities.append(switch)
+
+        add_devices_callback(DEVICES[config[CONF_HOST]].entities)
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, dispose)
+    return True
+
+
+def dispose(event):
+    """Close connections to Netio Devices."""
+    for _, value in DEVICES.items():
+        value.netio.stop()
+
+
+class NetioApiView(HomeAssistantView):
+    """WSGI handler class."""
+
+    url = URL_API_NETIO_EP
+    name = "api:netio"
+
+    def get(self, request, host):
+        """Request handler."""
+        data = request.args
+        states, consumptions, cumulated_consumptions, start_dates = \
+            [], [], [], []
+
+        for i in range(1, 5):
+            out = 'output%d' % i
+            states.append(data.get('%s_state' % out) == STATE_ON)
+            consumptions.append(float(data.get('%s_consumption' % out, 0)))
+            cumulated_consumptions.append(
+                float(data.get('%s_cumulatedConsumption' % out, 0)) / 1000)
+            start_dates.append(data.get('%s_consumptionStart' % out, ""))
+
+        _LOGGER.debug('%s: %s, %s, %s since %s', host, states,
+                      consumptions, cumulated_consumptions, start_dates)
+
+        ndev = DEVICES[host].netio
+        ndev.consumptions = consumptions
+        ndev.cumulated_consumptions = cumulated_consumptions
+        ndev.states = states
+        ndev.start_dates = start_dates
+
+        for dev in DEVICES[host].entities:
+            dev.update_ha_state()
+
+        return self.json(True)
+
+
+class NetioSwitch(SwitchDevice):
+    """Provide a netio linked switch."""
+
+    def __init__(self, netio, outlet, name):
+        """Defined to handle throttle."""
+        self._name = name
+        self.outlet = outlet
+        self.netio = netio
+
+        if self.netio.update.__name__ != 'wrapper':
+            self.netio.update = util.Throttle(MIN_TIME_BETWEEN_SCANS)(
```

Comment:
This is a hack…

I prefer you pass a throttled update as a parameter of the constructor.